Implement CSE in KernelFormatterHandler.ir_to_string

isuruf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116959

Since the string given here is used in inner_fn_str_len to check
if a kernel should be realized or not, reducing temp variables
will result in more complicated looking, but simple kernels
to not be realized.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

isuruf committed:
Reduce temp variables in KernelFormatterHandler.ir_to_string

Since the string given here is used in inner_fn_str_len to check
if a kernel should be realized or not, reducing temp variables
will result in more complicated looking, but simple kernels
to not be realized.

[ghstack-poisoned]

lezcano commented:
Actually, given that this increases the number of kernels that will be fused, could you please trigger a bench run to make sure it doesn't regress?

isuruf committed:
Update on "Implement CSE in KernelFormatterHandler.ir_to_string"


Since the string given here is used in inner_fn_str_len to check
if a kernel should be realized or not, reducing temp variables
will result in more complicated looking, but simple kernels
to not be realized.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

peterbell10 commented on torch/_inductor/virtualized.py:
```diff
@@ -152,9 +153,12 @@ def inner(*args, **kwargs):
             if name == "indirect_indexing":
                 return line
+            if line in self.seen:
+                return self.seen[line]
```
I'm not sure this is a good idea. Say for example, you have two IRs:
```
c = a + b
d = c * c
```
or 
```
d = (a + b) * (a + b)
```

These would be printed as identical by this printer, but they might not be treated the same by the rest of our analysis.

lezcano replied:
Note that this function is used in our crazy analysis where we stop fusion if the kernel is too large (lol)
```
    def has_exceeded_max_reads(self):
        return isinstance(self.data, Pointwise) and (
            self.num_reads() > config.realize_acc_reads_threshold
            or self.inner_fn_str_len() > config.realize_bytes_threshold
        )
```
I would assume that doing this sort of CSE here is not only sound but desirable.

peterbell10 replied:
Yes, but it would be better to make the heuristic better rather than changing the printing to make the existing heuristic less terrible.

isuruf replied:
I could implement a crude op counter instead of this, but we need CSE in either case.

lezcano replied:
This stack is looking into improving that heuristic. This is just one step in that direction.

peterbell10 replied:
I think an op counter with CSE makes much more sense, also note we already have an op counter that doesn't support CSE:

https://github.com/pytorch/pytorch/blob/7073dc604e220e1690dc21c77b52da1e63a534e1/torch/_inductor/dependencies.py#L298

It might be worth auditing places where that op counter is used to see if CSE makes more sense for all users.

lezcano replied:
I think that going from length of kernel to op count is changing one not-so-good heuristic for another. I think a better way forward would be to
1. See if we can remove all this heuristic altogether
2. If we see some actual regressions in some models, we should look at those models in particular, see which kernels in them were regressed, and put together a more hardware-based heuristic for them.

For point 2. @shunting314 put together some great tooling at https://github.com/pytorch/pytorch/pull/112450. See e.g. https://github.com/pytorch/pytorch/pull/100331#issuecomment-1666300394 or https://github.com/pytorch/pytorch/pull/100331#issuecomment-1716742135. You could touch base with him to discuss a good approach towards this sort of issues.

lezcano commented:
The CI is all red. Could you make sure the CI is green and then rebench? Weirdly enough, the failures did not clearly show up in the perf run.

isuruf committed:
Update on "Implement CSE in KernelFormatterHandler.ir_to_string"


Since the string given here is used in inner_fn_str_len to check
if a kernel should be realized or not, reducing temp variables
will result in more complicated looking, but simple kernels
to not be realized.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf commented:
Closing in favour of https://github.com/pytorch/pytorch/pull/117030
